### PR TITLE
Enable default compression for DataSet save().

### PR DESCRIPTION
**Pull request by crockpotveggies**
richard-hendricks

## What changes were proposed in this pull request?

The default signature for DataSet.save() now has compression enabled by default. 

Addresses https://github.com/deeplearning4j/nd4j/issues/1224.

## How was this patch tested?

Ran unit tests in addition to tested in a real-world project. Float16 was used for compression since it appears to be the best compression/lossless algorithm.

Weissman score is off the charts!

### CODE REVIEW

**Comment by crockpotveggies** [comment]
First question: do we want/need to update our interface API?

**Comment by agibsonccc** [comment]
@crockpotveggies yeah please do

**Comment by agibsonccc** [comment]
We *could* just dump a warning from save on that datatype?

**Comment by AlexDBlack** [comment]
I'd be fine with adding a one-off warning (once per JVM). "Default DataSet saving configuration is set to FP16. This may result in precision loss. Use .save(...) to specify compression codec, or set default compression codec using Y. Read Z for more details."

**Comment by crockpotveggies** [comment]
Hmmm, I'm getting some strange results when loading my presaved dataset w/ Parallel Wrapper, what's wrong here?

The score is also insanely large. On a single worker, it's 9.x as it should be.

```
[AsyncIterator thread] WARN org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer - Loading a data stream with type different from what is set globally. Expect precision loss
[AsyncIterator thread] WARN org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer - Loading a data stream with type different from what is set globally. Expect precision loss
[AsyncIterator thread] WARN org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer - Loading a data stream with type different from what is set globally. Expect precision loss
[AsyncIterator thread] WARN org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer - Loading a data stream with type different from what is set globally. Expect precision loss
[AsyncIterator thread] WARN org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer - Loading a data stream with type different from what is set globally. Expect precision loss
[AsyncIterator thread] WARN org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer - Loading a data stream with type different from what is set globally. Expect precision loss
[AsyncIterator thread] WARN org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer - Loading a data stream with type different from what is set globally. Expect precision loss
[AsyncIterator thread] WARN org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer - Loading a data stream with type different from what is set globally. Expect precision loss
[main] INFO org.deeplearning4j.parallelism.ParallelWrapper - Averaged score: 179947.4941971561
[AsyncIterator thread] WARN org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer - Loading a data stream with type different from what is set globally. Expect precision loss
[AsyncIterator thread] WARN org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer - Loading a data stream with type different from what is set globally. Expect precision loss
[AsyncIterator thread] WARN org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer - Loading a data stream with type different from what is set globally. Expect precision loss
[AsyncIterator thread] WARN org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer - Loading a data stream with type different from what is set globally. Expect precision loss
[AsyncIterator thread] WARN org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer - Loading a data stream with type different from what is set globally. Expect precision loss
[AsyncIterator thread] WARN org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer - Loading a data stream with type different from what is set globally. Expect precision loss
[AsyncIterator thread] WARN org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer - Loading a data stream with type different from what is set globally. Expect precision loss
```

**Comment by crockpotveggies** [comment]
@raver119 any ideas on the problem above? ^^^^

**Comment by raver119** [comment]
No, poke me in pm & give something to run.

**Comment by crockpotveggies** [comment]
Narrowed this down to incompatibility w/ PW+compression+workspaces. @raver119 will fix before merging this.

**Comment by raver119** [comment]
Making default save() call to save compressed FP16 datasets is terrible idea. It works ok only for images, for anything else, like nlp tasks - it's just disaster. Compression should be optional. I'm reverting default save() signature to work without compression.

**Comment by raver119** [comment]
All we need - to provide option to save compressed datasets. But enforcing lossy compression by default - is terrible UX. I can already see users popping up with questions :)

**Comment by raver119** [comment]
@crockpotveggies once you merge this PR, please merge this one: https://github.com/deeplearning4j/deeplearning4j/pull/3257

It adds decompression hook into ADSI/AMDSI

**Comment by AlexDBlack** [comment]
hm... so thinking about FP16 again: 6.1e-5 is smallest positive value and 2048 is the largest integer that can be represented exactly...
65504 is the largest value that can be represented at all. Given DataSets will often be saved pre-normalization, it's pretty much guaranteed that many users will hit this (and not even realize it). That's a pretty big downside for only a 2x space saving (it might be fine for many use cases - but users should probably have to make that decision for themselves)

**Comment by crockpotveggies** [comment]
Let's decide the "best" fast/lossless compression before merging this and/or abandoning this? The amount of space uncompressed data uses is fairly big. I think in real world situations for many big datasets, it might not be practical to use uncompressed saving. I'm happy to investigate snappy and lz4.

**Comment by raver119** [comment]
The only lossless compression that's implemented atm is gzip, and i doubt it's good thing perf-wise. But if we need something, we could make fast impl in native, i remember Adam included zlib in libnd4j anyway.

**Comment by crockpotveggies** [comment]
Performance on GZIP was *fairly bad* when I tested it. At least 10X than other methods. Yes can we look at a native impl?

**Comment by crockpotveggies** [comment]
What about Zstandard? https://code.facebook.com/posts/1658392934479273/smaller-and-faster-data-compression-with-zstandard/

**Comment by agibsonccc** [comment]
Just of note - I got rid of zlib it was causing too many problems. zlib was only used for saving/loading compressed numpy arrays which we weren't using anyways.

**Comment by crockpotveggies** [comment]
Okay going with lz4. @raver119 you were concerned that lz4 was lossy? But the algo claims it isn't?

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/DataSet.java [hunk 1]
```diff
@@ -442,10 +443,74 @@ public void load(File from) {
...
+            dos.flush();
+            dos.close();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
 
     @Override
     public void save(OutputStream to) {
+        saveUncompressed(to);
```
Comment (huitseeker) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/DataSet.java]: Should this really be the default ??

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/DataSet.java [hunk 2]
```diff
@@ -442,10 +443,74 @@ public void load(File from) {
         }
     }
 
+    public void save(OutputStream to, String compressionAlgorithm) {
```
Comment (huitseeker) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/DataSet.java]: This needs javadoc, not. on the value of `compressionAlgorithm`, please.

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/MultiDataSet.java [hunk 3]
```diff
@@ -674,6 +674,9 @@ public int hashCode() {
     public long getMemoryFootprint() {
         long reqMem = 0;
 
+        // we need to decompress all arrays before checking its size
```
Comment (huitseeker) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/MultiDataSet.java]: Grammar.

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/api/DataSet.java [hunk 4]
```diff
@@ -324,4 +324,9 @@
      * @return
      */
     long getMemoryFootprint();
+
+    /**
+     * Thia method checks, if dataset is compressed, and applies decompression to it
```
Comment (huitseeker) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/api/DataSet.java]: remove commas.

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/api/MultiDataSet.java [hunk 5]
```diff
@@ -172,4 +172,9 @@
 
     long getMemoryFootprint();
 
+    /**
+     * Thia method checks, if dataset is compressed, and applies decompression to it
```
Comment (huitseeker) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/dataset/api/MultiDataSet.java]: Ugh, commas.

**Comment by agibsonccc** [comment]
Closing due to inactivity.

**Pull request closed without merging**